component/status.go

```diff
@@ -0,0 +1,141 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package component // import "go.opentelemetry.io/collector/component"
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+type Status int32
+
+// Enumeration of possible component statuses
+const (
+	StatusNone Status = iota
+	StatusStarting
+	StatusOK
+	StatusRecoverableError
+	StatusPermanentError
+	StatusFatalError
+	StatusStopping
+	StatusStopped
+)
+
+// String returns a string representation of a Status
+func (s Status) String() string {
+	switch s {
+	case StatusStarting:
+		return "StatusStarting"
+	case StatusOK:
+		return "StatusOK"
+	case StatusRecoverableError:
+		return "StatusRecoverableError"
+	case StatusPermanentError:
+		return "StatusPermanentError"
+	case StatusFatalError:
+		return "StatusFatalError"
+	case StatusStopping:
+		return "StatusStopping"
+	case StatusStopped:
+		return "StatusStopped"
+	}
+	return "StatusNone"
+}
+
+// errorStatuses is a set of statuses that can have associated errors
+var errorStatuses = map[Status]struct{}{
+	StatusRecoverableError: {},
+	StatusPermanentError:   {},
+	StatusFatalError:       {},
+}
+
+// StatusEvent contains a status and timestamp, and can contain an error
+type StatusEvent struct {
+	status    Status
+	err       error
+	timestamp time.Time
+}
+
+// Status returns the Status (enum) associated with the StatusEvent
+func (ev *StatusEvent) Status() Status {
+	return ev.status
+}
+
+// Err returns the error associated with the StatusEvent.
+func (ev *StatusEvent) Err() error {
+	return ev.err
+}
+
+// Timestamp returns the timestamp associated with the StatusEvent
+func (ev *StatusEvent) Timestamp() time.Time {
+	return ev.timestamp
+}
+
+// StatusEventOption applies options to a StatusEvent.
+type StatusEventOption func(*StatusEvent) error
+
+// ErrStatusEventInvalidArgument indicates an invalid option was specified when creating a status
+// event. This will happen when using WithError for a non-error status.
+var ErrStatusEventInvalidArgument = errors.New("status event argument error")
+
+// WithError sets the error object of the StatusEvent. It is optional
+// and should only be applied to an Event of type ComponentError.
+func WithError(err error) StatusEventOption {
```

Comment:
How does this correspond to the different types of error statuses?

Reply:
Should we just have `NewRecoverableErrorEvent`, `NewPermanentErrorEvent`, `NewFatalErrorEvent`, each which require an error parameter?

Reply:
The comment is incorrect. WithError should be usable with `StatusRecoverableError`, `StatusPermanentError` and `StatusFatalError`. I prefer to keep the functional option if at all possible since due to the signature of report component status (e.g. `ReportComponentStatus(component.Status, ...component.StatusEventOption)`). Is it fine if I just fix the comment 😅?

Reply:
> due to the signature of report component status

I think the signature may be problematic. It just seems a bit too easy for the caller to use it incorrectly, e.g. report an error status without an error, or report a non-error status with an error. Incorrect usage may not be immediately apparent either because the cases where we report errors may happen very rarely. I'm worried that the effect of misusing it may not be noticed until we're in the exact situation where we really need it to work correctly.

It seems like there must be a better signature that can't be misused. Maybe it should just accept a `StatusEvent`, and we'll make it as easy as possible to construct events appropriately?

Reply:
I removed the functional options, replaced them with per event type constructors, and updated `ReportComponentStatus` to take a `StatusEvent`.